### ISSUE TITLE
Dissipative Particle Dynamics

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#1435]
There seems to be some interest in having support for DPD.  Most likely we would want to use the method described in http://arxiv.org/abs/cond-mat/0602523.  It would involve writing a new Integrator that integrates each pairwise distance separately.

**Comment by golobor** [comment]
Dear Peter,
out of curiosity, do you still plan to implement DPD in OpenMM?
The reason I'm asking is that I found that DPD can lead to some crazy speed-ups (a preliminary estimation is 25-100x) of coarse-grained high-density simulations of chromosomes in Hoomd-blue. Our main tool of choice is OpenMM and I wish I could implement custom integrators in OpenMM, but I suspect that this may be above my skill level (I know *some* C++ and understand the basics of the integration of motion equations, but all of this is fairly rudimentary).
Anton.

**Comment by giadefa** [comment]
HI Anton,
I worked with DPD for my years. Can you clarify speed-ups compared to what?
DPD is ultimately an expensive thermostat which preserves hydrodynamics
plus a conversative force field which is independent from DPD.

**Comment by golobor** [comment]
Hi Gianni,

many thanks for answering! 
This is an excellent question, indeed! 

As I've mentioned, I use MD to perform coarse-grained simulations of soft biological matter, predominantly chromosomes. For this comparison, I simulated a single long polymer of 1e4 particles (mass = 1.0 [units of mass]), connected via not-so-stiff harmonic bonds (k=20 [force/distance], r_eq = 1.0 [distance]), in periodic boundary conditions at a high density (the specific volume per particle is 1.25^3, while the particle diameter is 1.0). The comparison is the following:

(L):
- a langevin thermostat (heavily underdamped, gamma=0.01 [force/velocity], kT = 1.0 [energy]) (https://hoomd-blue.readthedocs.io/en/latest/module-md-methods.html#hoomd.md.methods.Langevin)
- a DPD conservative repulsive field, r_max = 1.0 [distance], peak force = 0.5 [force] 
- Velocity-Verlet integration with dt = 0.005 [time] 

(D):
- a DPD force/thermostat (gamma=10 or 1.0 [force/velocity], kT = 1.0 [energy], r_max = 1.0 [distance]) (https://hoomd-blue.readthedocs.io/en/latest/module-md-pair.html#hoomd.md.pair.DPD) 
- a DPD conservative repulsive field, r_max = 1.0 [distance], peak force = 0.5 [force] 
- Velocity-Verlet integration with dt = 0.005 [time] 

The two simulations seem to be fairly comparable, except for the difference in viscosity. In the case of Langevin simulations, underdamping greatly speeds up equilibration, so I picked a relatively low value of viscosity. In the case of DPD, viscosity does not seem to affect the speed of equilibration so much. (both can be tested systematically, if needed)

The key result is that the DPD simulations (D) show much faster diffusion per step, comparing to the Langevin ones (L). 
In DPD, the average displacement of a particle is ~5.0 [distance] per 10^4 steps, while in Langevin it is around ~1.0 [distance]. Given that the polymer dynamics is subdiffusive (<dx> ~ t ^ 0.25), this can potentially mean massive gains in performance (up to 5^4, though it may be an overestimate). I have not measured GPU time per step, but it doesn't seem to vary by more than 2x.

This is further supported by one more simulation, where I managed to equilibrate a single-polymer system of 1e6 particles over 8 hrs using DPD on a single GPU, while it wouldn't equilibrate in underdamped Langevin simulations over 8 days. 


Again, this of course is still somewhat anecdotal, but, if you think it could be potentially interesting and/or useful to you, I can wrap these studies up into a more rigorous document. 
I am no specialist in DPD (in fact, I really understood it only a couple of weeks ago), but my intuitive interpretation of this speed-up is that the dissipative interactions suppress oscillations between adjacent particles in dense phases and allow spending computational time on simulations of correlated motions of whole clusters/regions of adjacent particles. 

Alternatively, I got something horribly wrong in my setup of the Langevin system, which could still be extremely useful for me to understand.

I'd be happy to provide any further info/simulation files/perform further simulations/etc.
Anton.

**Comment by giadefa** [comment]
Hydrodynamics will contribute to long-range correlations which will speed
up diffusion of such large systems. Usually this allows you to see more
physics, rather than just being a speedup problem. I doubt that the speedup
is of the orders that you mentioned.

Why do you want to use openMM if you are already simulating what you want
with hoomd?

g

**Comment by golobor** [comment]
Yes, I agree that implementing hydrodynamics is more about new physics, rather than a mere speed-up. In my message above, I formulated DPD's advantage rather narrowly, only b/c the specific scientific problem I was working on (simulation of compacted chromosomes) is an equilibrium problem (*most likely) and my research was severely limited by the speed of equilibration for many years now. It is very exciting for me to finally see a technical solution for this. 

Re: speedup being in the orders that I've mentioned - I do suspect that, for this specific problem, my estimates are not terribly off. Again, it involves the relaxation of very long polymers, with an essentially flat energy landscape (except for collisions) - so equilibration is mostly driven by diffusion (also, polymers are sub-diffusive with a rather nasty length scaling). We've been using Langevin for years now and, as I strongly suspect, have fully optimized its equilibration speed. So, the 5x higher displacement in DPD for the same number of steps is really game-changing for us.

Re: openmm vs hoomd, we have been using OpenMM for a decade now and created several packages for coarse-grained simulations (https://github.com/open2c/polychrom , https://github.com/golobor/wiggin) and it really helped us to model chromatin folding (eg. https://www.cell.com/cell-reports/comments/S2211-1247(16)30530-7 , https://pubmed.ncbi.nlm.nih.gov/24200812/ , https://pubmed.ncbi.nlm.nih.gov/29348367/ , etc). In fact, I only learned hoomd to test the performance of the DPD thermostat. While I liked hoomd much more than I expected, I still found it rather limiting b/c of the way it implements forces. The ability to define forces with custom expressions in OpenMM is an extremely important factor for the development of coarse-grained simulations, where the exact potentials are neither known, nor important  (as a note, I do know that hoomd accepts tabulated functions, but this functionality seems limited too). Moreover, some of the simulations of phase separation of complex biological molecules that we're currently developing will involve CustomManyParticleForce or CustomHbondForce and do not seem possible with hoomd.

**Comment by giadefa** [comment]
hi,
thanks for the clarifications.

I expect that we will be implementing more CG stuff in OpenMM, because we
are working on it. A DPD thermostat could be one, but I don't have any
definite plan at the moment.

**Comment by golobor** [comment]
thank you!

**Comment by peastman** [comment]
Thanks for your comments!  Features mostly get prioritized based on user interest.  In the six years since I opened this issue,  no one ever posted to say they were interested in it.  That made it an obvious candidate to keep getting pushed off to a later release. :)  Knowing you're interested in it increases its priority!

**Comment by golobor** [comment]
continuing a discussion at https://github.com/openmm/openmm/issues/3807
An alternative to DPD could be the Loewe-Andersen thermostat (https://aip.scitation.org/doi/10.1063/1.2198824). This thermostat is similar to Andersen's, except it acts on randomly chosen pairs of colliding particles and randomizes the radial components of their velocities.
<img width="660" alt="image" src="https://user-images.githubusercontent.com/912906/196375584-a67be053-b20d-4c25-8881-cb3662f4866d.png">
It seems, however, equally challenging to implement in OpenMM, at least at first glance.

**Comment by svarner9** [comment]
I am wondering if implementing the DPD thermostat has gotten any more traction internally? I love using OpenMM due to the incredible speeds (especially in v8.1) on a single GPU and the simple python interface. Lately, however, I am running simulations where I expect hydrodynamics to be important, which requires the DPD thermostat.

Namely, I am simulating morphological changes in dense polymer systems. Currently I am just using the Langevin middle integrator with a custom pair potential that mimics the DPD conservative force. This works for simulating soft-matter systems over long length and time scales, but does not preserve hydrodynamics.

Based on the previous discussion in this thread and in #3807 , it seems that implementing the DPD thermostat is highly nontrivial and would not fit the mold of any integrators currently available in OpenMM, including the custom integrator. If it is not a priority, is there any initial advice you could provide in case I try my hand at implementing it over the next several months? (e.g., location of the most relevant source code, examples where neighborlists are used, examples where random numbers are generated).

Thank you so much for your time and consideration, and I look forward to hearing your response!

**Comment by giadefa** [comment]
You could start from here:
https://github.com/compsciencelab/mydpd

**Comment by peastman** [comment]
It's been a long time since I looked at it.  If I recall, DPD requires a more complicated integration scheme but interfaces with the rest of the code much like any other integrator.  For example, it doesn't require access to the forces from individual nonbonded interactions, just the total force on each particle.  Is that correct?  If so, it should be fairly straightforward to implement it as a plugin.

If you're interested in doing that, the place to start is the [developer guide](http://docs.openmm.org/latest/developerguide/).  It describes the architecture for plugins.  Then look at how the existing integrators are implemented, and also the [example plugin](https://github.com/openmm/openmmexampleplugin).  It implements a force rather than an integrator, but most of the structure is the same.

**Comment by peastman** [comment]
I looked into this a bit more.  Equation 1 in the paper above seems to assume the forces are strictly pairwise (no angle or torsion terms) and finite ranged (no PME).  Are those really requirements?

It seems like this method doesn't work with general potential functions.  What sort of an API would make sense for it?  Does the integrator need to provide its own mechanism for defining the potential using only a restricted set of primitives?  What would you imagine the API looking like?

**Comment by giadefa** [comment]
https://arxiv.org/abs/cond-mat/0602523

**Comment by peastman** [comment]
Yes, that's the paper I was referring to.  Equation 1 integrates the positions based on a pairwise interaction force with no way to incorporate a force that isn't pairwise.

**Comment by giadefa** [comment]
that's just the DPD thermostat, then you can add a conservative force as
well in the Trotter expansion.

This is better:
https://arxiv.org/pdf/physics/0502064

**Comment by giadefa** [comment]
Eq. 16 is MD with a DPD thermostat.

**Comment by peastman** [comment]
The only force in eq. 16 is $F_c(r_{kl})$, a pairwise interaction force depending only on the distance between two particles.  How would you modify it to allow arbitrary potential functions?

**Comment by peastman** [comment]
If you assume $a_{kl}$ is constant for all interactions, you can pull that term out of the sum and it becomes just the total force on the particle.  It isn't clear to me what $a_{kl}$ is supposed to represent.  What is its function, and what values would it normally have?

**Comment by giadefa** [comment]
a_kl is 1 in MD.

In DPD simulations they often use a_kl to represent hydrophobic,
hydrophilic interactions between particles and a very simple F_c.
Just a bad notation, you can set a_kl=1 for all k and l

**Comment by svarner9** [comment]
I want to start by saying thank you for continuing the discussion and showing an interest in this issue! DPD in OpenMM would be a wonderful advancement for the soft matter community. I am excited to know that this is gaining interest once again. Below are some of my thoughts and observations after using DPD in a handful of other simulation packages.

F_c is the conservative force that was originally used with DPD. In the [Trotter algorithm paper](https://arxiv.org/pdf/physics/0502064), they mention below Equation 17 that F_c(r) usually equals w(r), which is why they have pulled the a_ij out. As Gianni mentioned, it is just bad notation on their part. It is just thought of as the interaction potential and does not need to be used in combination with the stochastic forces. In general, any pair potential can be used with the DPD stochastic forces (thermostat) (e.g. LJ, Gaussian, Morse, etc). You can even include smeared electrostatics on top as well as is done with the [Slater potential](https://docs.lammps.org/pair_dpd_coul_slater_long.html).

In [LAMMPS](https://docs.lammps.org/pair_dpd.html), you have the option to use the DPD conservative force with the DPD thermostat attached, but you also have the option to use the DPD thermostat with any other conservative force you may want. Similarly, in [MATILDA.FT](https://github.com/rar-ensemble/MATILDA.FT/blob/main/src/Extraforce_dpd.cu) you can use the DPD thermostat with any pair potential. The same is the case in [HOOMD-Blue](https://hoomd-blue.readthedocs.io/en/latest/hoomd/md/pair/dpd.html).

After thinking about it further, I think it actually would be much simpler to instead think of the DPD thermostat as a custom non-bonded force that includes only the random force and the drag force, with the additional caveat that it depends on pairwise distance **vectors** and velocities (instead of just positions), and also requires random number generation. As mentioned in [another issue](https://github.com/openmm/openmm/issues/3807#issuecomment-1281208653), it cannot be done through the Python interface since CNBF in OpenMM doesn't allow for vector distances (correct me if I am wrong), however, a new class could be created with a slightly altered C++/CUDA backend that does have access to vector distances and velocities.

If this can be done, then the stochastic force can be used on top of whatever conservative force is desired (usually the DPD conservative force). My colleague has already implemented the [DPD conservative force](https://mdcraft.readthedocs.io/en/latest/api/mdcraft.openmm.pair.dpd.html#mdcraft.openmm.pair.dpd) as a custom non-bonded force in OpenMM, which takes all the a_ii and a_ij as parameters, as well as r_cut. The implementation of the DPD stochastic force (random and drag) should only rely on the noise parameter (sigma, or gamma=0.5*sigma^2 / k_B T), the temperature, and the cutoff (usually the same as the conservative force cutoff). In general, you can have a type-dependent noise parameter (i.e. different gamma_ii and gamma_ij).

In any case, the actual integration is still traditionally done through an NVE integrator like VV or the better [MVV algorithm](https://pubs.aip.org/aip/jcp/article/107/11/4423/478600/Dissipative-particle-dynamics-Bridging-the-gap) (Equation 9) which takes account for the fact that the force depends on the velocity. You can see this is the case in the LAMMPS docs and HOOMD docs (both linked above). The Trotter scheme appears to be more complicated and more accurate than traditional VV, but utilizing this scheme would be going down the DPD as an integrator route, as opposed to the DPD as a force route, as the DPD stochastic forces are _baked into_ the integrator. The Trotter algorithm presupposes the form of the stochastic forces and then derives the result of an integration step from there. In almost all modern software packages, DPD is implemented as a force, and utilized with a regular VV or MVV integrator, which I believe would be a simpler solution as of right now.

[This paper](https://doi.org/10.1016/j.jcp.2011.05.021) from the Glotzer group may be more useful since it gives the pseudocode for the algorithm on GPUs to optimize the random number generation aspect. As they say, "The DPD pair-wise force can be simply incorporated into the existing pair-wise force summation kernel". With all of this in mind, I think perhaps it may be simpler than expected! The main tasks would be incorporating vector distances, velocities and random number generation into a modified version of the custom nonbonded force class. Special care has to be taken in the CUDA kernel implementation to ensure that the two threads which compute F_ij and F_ji utilize the same random number. This is described in Algorithm 4 of the Glotzer paper, which utilizes a so-called _pK-pT_ scheme.

Best,
Sam

**Comment by peastman** [comment]
I admit to being kind of confused about DPD.  Papers on it tend to describe it in ways that don't make sense to me.  For example, it's usually described as being an alternative to molecular dynamics, but it looks to me like it *is* molecular dynamics, just with a particular set of forces.  And it's always presented as involving only pairwise interactions, but I can't find any reason for that limitation.

Following Gianni's advice that $a_{kl}=1$, we can pull the first term out of the sum.  Equation 1 in http://arxiv.org/abs/cond-mat/0602523 then becomes

$dp_k = F_k + \sum_{l \neq k}^{N} e_{kl} \[ -\frac{\gamma}{m} \omega^2(r_{kl})(e_{kl} \cdot p_{kl}) dt + \sqrt{2 \gamma k_BT_0} \omega(r_{kl}) dW_{kl}^t \]$

where $F_k$ is the total conservative force on particle $k$.  That looks like an ordinary Langevin equation, except that the friction and noise forces are applied to inter-particle displacements instead of Cartesian coordinates.  That makes it more expensive to compute, but shouldn't change anything fundamental.  We should be able to compute it with the same LF-Middle discretization we use for other Langevin integrators.

But it was assumed from the start that $F_k$ includes only pairwise terms.  Does that really matter?  What if it also includes a three particle angle term?  Will it still work, or does that somehow break the method?

**Comment by svarner9** [comment]
First off, I apologize if my suggestions are very naive and infeasible, as I am not super familiar with how the OpenMM backend works, or what is allows.

I still think it might be easier to just treat the stochastic part of DPD as a force that can be overlayed on top of any other forces in the system. Then the system is just integrated with a traditional NVE integrator. It makes sense to do it this way since those forces are all pairwise, they can be computed at the same time as you compute all of the other non-bonded forces in the system (which are also pairwise). It is my understanding that none of the current integrators consider pairwise distance, velocities, and also don't have access to any neighborlists. The fact that the DPD stochastic forces are pairwise and can have different friction parameters between different types of particles naturally makes it suitable as a pairwise nonbonded force in MD codes. Forgive me though as this may not be the case in OpenMM.

Note that it is not physical to use the DPD stochastic forces and then also use a thermostating method like a Langevin thermostat. If the DPD forces are added as a nonbonded force, then one must use an NVE integrator. On the other hand, you can implement DPD as an integrator, but I see a few problems with this. One issue is that the random and drag forces wouldn't be included in the virial for each particle, and so I am not sure how that methodology would interact with a barostat. In general, only the conservative force is included for the potential energy calculation, and so that part would not be an issue.

If the conservative force in your system is from a multibody potential, then I am not 100% sure how the stochastic forces need to be treated. As an example, there is a multibody version of DPD ([mDPD](https://docs.lammps.org/pair_mesodpd.html)), and there it seems that the stochastic forces are still simply pairwise additive and they have the same form as classic DPD.

These confusions may be a result of me trying to adapt OpenMM for something that it isn't really intended for. OpenMM is really designed for atomistic and biomolecular simulations, not highly coarse-grained mesoscopic simulations. It is also possible that the force computation backend in OpenMM is inflexible to the kinds of changes that I am proposing such as including a kernel that takes quantities r1, r2, v1, v2, and R (a random number) and computes d=r1-r2 (the displacement vector), and then computes the stochastic forces of DPD. I am not familiar enough with the C++/CUDA backend to know if this is feasible. Perhaps trying to adapt OpenMM to compute these additional forces will cancel out any speed benefits that one would hope for when switching to OpenMM.

Thank you!

**Comment by peastman** [comment]
The friction and noise could be implemented either as a force or an integrator.  There are theoretical reasons to prefer the latter.  Modern methods for integrating Langevin equations apply the noise and friction at a different point in the step from other forces.  This leads to a cancellation of errors and higher accuracy.

Putting that question aside, though, if we integrate the equation shown above with $F_k$ being an arbitrary conservative force (not necessarily all pairwise), will that produce theoretically sound results?

**Comment by bbye98** [comment]
Apologies for jumping into this conversation, but I happened to see that this thread is active again.

I believe that the DPD conservative force can include non-pairwise interactions, like angle and dihedral potentials that involve more than two particles. For example, [this LAMMPS tutorial](https://www.lammps.org/workshops/Aug11/Sirk/lammps_pres.pdf) uses a harmonic angle potential, and [this GROMACS tutorial](https://dl-sdg.github.io/RESOURCES/TUTORIALS/dlm_8.html) mentions the possibility of adding on angle and dihedral potentials.

**Comment by giadefa** [comment]
The only important thing is that F_c is conservative and it does not depend
on P, t directly. So anything that we use in MD is fine.

The thermostat needs to be calculated in a pairwise manner because it needs
to preserve local momentum.

In the mydpd repo above,  you find the best Trotter integrator for it. It's
in C, but it should not be hard to move it to CUDA inside openMM.

**Comment by peastman** [comment]
Thanks!  That's what I suspected.

**Comment by svarner9** [comment]
> The friction and noise could be implemented either as a force or an integrator. There are theoretical reasons to prefer the latter. Modern methods for integrating Langevin equations apply the noise and friction at a different point in the step from other forces. This leads to a cancellation of errors and higher accuracy.
> 
> Putting that question aside, though, if we integrate the equation shown above with F k being an arbitrary conservative force (not necessarily all pairwise), will that produce theoretically sound results?

Is it okay to apply the stochastic forces at a different point in the step in the case of DPD? Only wondering because of the need to conserve momentum.

**Comment by peastman** [comment]
Yes.  Each piece individually conserves momentum, so it's conserved whether you apply them at the same time or at different times.

**Comment by peastman** [comment]
Two more questions.

The paper says a "typical" choice is $\omega(r)=1-r/r_c$.  Can we assume that form?  Or do we need to support other forms as well?

The equation for $dp_k$ sums over neighbors $l$.  The expression that gets summed includes the mass $m$.  Is that meant to be $m_k$?  Or the reduced mass of atoms $k$ and $l$?

**Comment by giadefa** [comment]
$\omega$, I don't remember why that was a good choice and if there are
others, I would think that it probably does not matter too much, it's just
a way to have a limited range.

For the mass, again, I am not sure, but I think that it's simplified by the
fact that the mass is the same for all particles here, what you probably
want is to remove the mass in the term and use
e_kl · v_kl

but again I am not 100% sure.

**Comment by svarner9** [comment]
$$\omega(r)=1-r/r_c$$ is the only weight function that is really used. I think you can assume that form and do not need to include support for other forms. The only requirement is that $$\omega_D=\omega_R^2$$ and therefore also $$\sigma^2=2\gamma k_B T$$ in order to satisfy the fluctuation dissipation theorem. The main functionality of the $$\omega(r)$$ functions is to set the range of the random and drag forces.

Also note that in general, one should be able to specify a different $$\gamma_{ij}$$ and a different $$r_c$$ for each combination of types.

As for the mass, in the original [Espanol and Warren 1995](https://iopscience.iop.org/article/10.1209/0295-5075/30/4/001) work, they only have the mass in the first equation below to update the position.

![Image](https://github.com/user-attachments/assets/6dcad53e-265c-4dad-a846-160e377c13b5)

**Comment by giadefa** [comment]
so, ok, what I said is correct

**Comment by peastman** [comment]
> Also note that in general, one should be able to specify a different $\gamma_{ij}$ and a different $r_c$ for each combination of types.

How important is that?  If it required a single cutoff and/or a single friction coefficient, would that be a problem, or not a big deal?

If it's important, how would you want to specify them?  Describe what you would want the API to look like.

**Comment by svarner9** [comment]
Hm that is a tough question. Personally, I have not used different $$\gamma_{ij}$$, but I can see where it would be useful to simulate mixtures that have differing friction between components. In addition, the $$\gamma$$ parameter can be directly related to the self-diffusion coefficient of the species, as shown in [Groot and Warren 1997](https://pubs.aip.org/aip/jcp/article/107/11/4423/478600/Dissipative-particle-dynamics-Bridging-the-gap), $$D=45k_B T / 2\pi \gamma \rho r_c^3$$, and therefore if you have a mixture of components that are expected to have very different self-diffusion coefficients, then you would need the ability to have different $$\gamma$$ and/or $$r_c$$, since $$\rho$$ is obviously fixed throughout the simulation box. In addition, people have used DPD to simulate colloidal particles in a small molecule solvent or polymer matrix, and in that case the $$r_c$$ for colloidal particles is much larger than that for the solvent or monomers in the system. I'm not an expert by any means, just a user who focuses on polymeric systems, so I would love to hear from others as well.

I think it would not be a huge deal if you did not allow for different $$\gamma$$ or $$r_c$$ since the majority of systems utilize fixed values for these. However, packages like LAMMPS and HOOMD do allow for these parameters to vary for different type-pairs, so OpenMM would remain relatively limited in that case.

For the API, assuming you go the integrator route, in an ideal world it may look like what I have written below. Again, I am not an expert, just a frequent user. I would love to hear the opinions of those closer to development and other DPD users as well.
1) Build system and topology as you always do
2) Specify non-bonded and bonded interactions as normal
3) Set integrator to DPD with a temperature ($$T$$), global cutoff ($$r_c$$), and random seed
4) Specify pairwise $$\gamma_{ij}$$ through a tabular function
5) Optionally specify $$r_c$$ through a tabular function, with the global cutoff being the default

Note that in the above, I don't mention the $$a_{ij}$$ parameters since we will assume the user has specified the conservative force in their non-bonded interactions.

Here is an example from the HOOMD documentation:

```ruby
nl = nlist.Cell()
dpd = pair.DPD(nlist=nl, kT=1.0, default_r_cut=1.0)
dpd.params[('A', 'A')] = dict(A=25.0, gamma=4.5)
dpd.params[('A', 'B')] = dict(A=40.0, gamma=4.5)
dpd.params[('B', 'B')] = dict(A=25.0, gamma=4.5)
dpd.params[(['A', 'B'], ['C', 'D'])] = dict(A=40.0, gamma=4.5)
```

**Comment by peastman** [comment]
That HOOMD code uses the same cutoff (1.0) and friction (4.5) for all pairs.  Do you know of any publications that vary them?  I just looked through some review articles, but didn't see any mention of anyone doing that.

**Comment by bbye98** [comment]
The $r_\mathrm{cutoff}$ can be thought of as an effective particle size. When you have systems of two types of particles with different particle sizes, you can used the "mixed" $r_{\mathrm{cutoff},~ij}$ to control how close particles from species $i$ and $j$ can get to each other.

I'm not too sure about $\gamma$, but given that LAMMPS and HOOMD-blue allow them to be specified for each pair type, that is likely the correct approach.

**Comment by peastman** [comment]
Could they be specified per-particle and then computed with a combining rule for each pair?

**Comment by bbye98** [comment]
This is the tricky part--there's no well-defined mixing rule for any of these term. We could go the route of using geometric mixing for $A_{ij}$ and averaging $r_{\mathrm{cutoff},~ij}$ like for $\epsilon_{ij}$ and $\sigma_{ij}$ in LJ, but I'm not sure what to do with $\gamma_{ij}$.

**Comment by peastman** [comment]
Don't worry about $A_{ij}$.  The potential will be defined with regular forces.  CustomNonbondedForce should be able to handle just about anything people want to do.

How about this as an API.  The constructor for DPDIntegrator will take `defaultCutoff` and `defaultFriction` arguments.  If you just want a single value for each one, that's all you need to specify.  Those values will be used for all particle pairs.

If you need different values for different pairs, you can call the methods `setParticleType(particleIndex, typeIndex)` and `addTypePair(typeIndex1, typeIndex2, cutoff, friction)`.  By default every particle has type 0.  If you add a type pair, any interaction between two particles of those types uses the specified parameters instead of the default ones.

**Comment by bbye98** [comment]
I personally think this is a brilliant solution! This would make the DPDIntegrator functionally identical to the implementations in other simulation packages..

**Comment by svarner9** [comment]
This sounds great! Only thing I would say is that usually $$\gamma$$ (or $$\sigma$$) is referred to as the noise parameter or noise level, since it sets both the amplitude of the drag force as well as the variance of the random force. That is just semantics though.

**Comment by peastman** [comment]
It sounds like we have a plan then.  Can I count on you to help test the implementation once it's written?

**Comment by svarner9** [comment]
Yes for sure! Thank you!

**Comment by peastman** [comment]
I have a first draft of the API, if you want to look it over.

https://github.com/peastman/openmm/blob/dpd/openmmapi/include/openmm/DPDIntegrator.h

Next is to write the reference implementation.

**Comment by svarner9** [comment]
Looks good!

**Comment by peastman** [comment]
I have the reference implementation written.  The code is in https://github.com/peastman/openmm/tree/dpd if you want to try it out.  It should work for both the Reference and CPU platforms, though I also plan to write a separate implementation for the CPU platform that will be faster.

**Comment by peastman** [comment]
Do you have any suggestions for how I can test it?  The current test cases are mostly just the same ones as for LangevinIntegrator: that it gives the right temperature, that it respects constraints, etc.  The only DPD specific tests are to check that it conserves momentum and that type pairs are interpreted correctly.  What are other ways to verify it's working correctly?

**Comment by peastman** [comment]
I'm now working on the GPU implementation.  I need to decide how to structure the kernels.  The most efficient way of doing that will depend on what a typical system is like.  Can you provide advice on the following?

1. What is a typical size for a system simulated with DPD?  What matters is the total number of particles.  If a typical simulation includes a few thousand particles, that will call for a different algorithm than if a typical simulation includes a million particles.
2. What is the typical density of particles, measured in terms of the cutoff distance?  If a typical particle has about ten neighbors within the cutoff distance, that calls for a different algorithm than if a typical particle has 1000 neighbors within the cutoff distance.

Regardless of how I write the kernel, it will work for a wide range of sizes and densities.  But if I know what to optimize for, I can make it as fast as possible in the typical cases.

**Comment by giadefa** [comment]
one use case is to use it as a thermostat in normal MD simulations.

This allows to study of the hydrodynamics pressure waves and other fluid
phenomena in realistic liquids.

e.g. https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.97.134501

g

**Comment by peastman** [comment]
What is the definition of reduced number density?

**Comment by bbye98** [comment]
> What is the definition of reduced number density?

It's just rho*r_cut^3, or the number density made dimensionless by multiplying by roughly the size of a particle.

Apologies for no LaTeX--on GitHub mobile.

**Comment by peastman** [comment]
So in a typical sphere of radius $r_c$, there will be about three particles?

**Comment by peastman** [comment]
Or rather, a typical sphere of volume $r_c^3$.

**Comment by bbye98** [comment]
@peastman Yes! Another way of thinking about it is that there are 3 DPD particles per unit volume, or that each DPD particle has 3 neighbors within r_c.

**Comment by peastman** [comment]
Thanks!  I'll optimize for the case of lots of particles, but few neighbors for each particle.

**Comment by peastman** [comment]
The implementation is in #4799.  It's ready to test, so please try it out and tell me how it works!

Also as noted above, any suggestions for more test cases I could add would be great.

**Comment by svarner9** [comment]
Hello! Wow it is great to see how much progress has been made on this! I am sorry I have been MIA for a bit, but I am back and ready to test things out!

Just a small comment. The reduced density is 3, and so the number of particles within a shell of radius r_c around a tagged particle is roughly $$n\approx\rho V-1\approx 3*4\pi/3-1\approx 11$$. Sometimes people even go to higher densities like $$\rho=5$$, which would correspond to $$n\approx 20$$. I am not sure if these differences are large enough to change your optimization, but I thought I should point it out just in case.

A possible test to verify that the dynamics are correct is to compute the diffusion coefficient, kinematic viscosity, and hence the Schmidt number, and verify that they match with the original Groot and Warren work. They found that for $$\gamma=6.75$$ (i.e. $$\sigma=3.675$$), $$\rho=3$$, and $$r_c=1.0$$, the dynamic properties were measured as $$D=0.306$$, $$\nu=0.305$$, and therefore $$Sc=1.00$$.

When the [DPD conservative force](https://mdcraft.readthedocs.io/en/latest/api/mdcraft.openmm.pair.dpd.html#mdcraft.openmm.pair.dpd) is used, we can verify that the equation of state is obeyed, $$P=\rho k_B T + 0.101a\rho^2$$ where $$a$$ is the repulsion strength in the conservative potential $$U(r)=\frac{1}{2}ar_c\left(1-\frac{r}{r_c}\right)^2$$ for $$r<r_c$$ and $$U(r)=0$$ otherwise, for a single component fluid.

I will continue to look for other tests and let you know if I come up with anything. In addition, I will compare results I have obtained in LAMMPS with results in OpenMM to make sure nothing is different.

**Comment by svarner9** [comment]
Alright I have been running some basic DPD simulations and I have some initial results, and some initial thoughts/questions.

I looked into lines 84-92 in `dpd.cc` and Eq 10 from the [Trotter paper](https://arxiv.org/abs/cond-mat/0602523) to try and match the OpenMM parameters with the traditional DPD convention.

```ruby
real mass1 = (vel1.w == 0 ? 0 : 1/vel1.w);
real mass2 = (vel2.w == 0 ? 0 : 1/vel2.w);
real m = mass1*mass2/(mass1+mass2);
real omega = 1-(r/cutoff);
real vscale = EXP(-dt*friction*omega*omega);
real noisescale = SQRT(1-vscale*vscale);
real3 dir = delta*invR;
real3 v = make_real3(vel2.x-vel1.x, vel2.y-vel1.y, vel2.z-vel1.z);
real dv = (1-vscale)*dot(v, dir) + noisescale*SQRT(kT/m)*getRandomNormal(random);
```

In the paper the exponential is $$\exp\left(-2\Delta t m \omega^2 / \gamma\right)$$, where $$\gamma$$ is the traditional DPD noise parameter.

In the code, $$m$$ will almost always be 1/2, which I assume is why $$vscale$$ does not need a factor of $$2$$ in the exponential? In addition, $$vscale$$ has $$\text{friction}$$ in the exponential instead of $$\gamma^{-1}$$. Thus, I assume the friction specified in the DPD integrator is related to the traditional DPD noise parameter via the relationship, $$\gamma = 1/\text{friction}$$.

I am probably missing something, but in the paper, the first term in $$dv$$ appears to begin with $$(vscale-1)$$, whereas the code uses $$(1-vscale)$$. Does the sign get flipped somewhere that I am not aware of? Does this have to do with the direction convention ($$k-l$$, versus $$l-k$$)?

In almost every case, one will set $$k_B=T=1$$, and therefore $$k_B T=1$$. In OpenMM, I don't think this is possible, so I just have to be careful to propagate the correct energy and time scales through everything. In order to do this, I set the following parameters:

1. $$M_W$$ = 0.001 kg/mol
2. $$r_c$$ = 1 nm
3. $$T$$ = 300 K

This naturally gives the following scales:

1. $$\epsilon$$ = energy = $$k_B T$$ = 2.49 kJ/mol
2. $$\sigma$$ = length = $$r_c$$ = 1 nm
3. $$\tau$$ = time = $$\sqrt{M_W r_c^2 / k_B T}$$ = 0.63 ps

Thus, it would be my understanding that I can set the timestep and friction in units of the timescale $$\tau$$ in order to make direct comparisons to the typical DPD conventions. Thus I set the following:

1. $$\Delta t=0.04\tau$$
2. $$\text{friction}=(\gamma\tau)^{-1}$$

Thus, in the exponential factors, the timescale $$\tau$$ will cancel identically. On the other hand, I am not 100% confident if this also takes care of the $$\sqrt{k_B T / m}$$ factor in the second term of the $$dv$$ expression.

For my test simulations I utilized the [DPD conservative force](https://mdcraft.readthedocs.io/en/latest/api/mdcraft.openmm.pair.dpd.html#mdcraft.openmm.pair.dpd) with a single component with $$A_{ii}=a=25$$, $$r_c=1$$, and a density of $$\rho=3.0$$. I tried various different values of $$\delta t = \Delta t / \tau$$ and $$\gamma = (\tau*\text{friction})^{-1}$$ within typical ranges.

For example, the recommended starting point for DPD simulations is $$\gamma=4.5$$ and $$\delta t\approx 0.04$$. When I run this simulation in OpenMM, the target temperature of 300 K is reached quickly and then fluctuates around the target by <1%. However, randomly after running for some time I get a NaN energy error. I tried running the simulation many many times and each time it failed at different points with NaN energy. There were no obvious reasons for this to occur given that the temperature remained pretty tight around the target of 300 K. Also keep in mind that these parameters are very standard for DPD simulations, so the simulation is expected to be very very stable. Thus, either the integration scheme is not so stable, there is some small error somewhere, or the way that I have mapped the parameters is not correct.

I also ran another case with a much smaller noise, $$\gamma=0.005$$, and timestep, $$\delta t = 0.015$$. I have run simulations with these parameters in LAMMPS, and they have been very stable around the target temperature with <0.1% temperature variation at equilibrium. On the other hand, with the OpenMM implementation the temperature fluctuates around 320 K for a while before eventually failing with a NaN energy.

I am hoping that there is just something wrong with the way I am mapping the parameters between the OpenMM convention and the traditional reduced unit convention used in other software for coarse-grained simulations.

I have uploaded the simulation file to a [repo](https://github.com/svarner9/OpenMM-DPD-Test) that I am using to test the DPD implementation. Note that it requires the use of the [MDCraft](https://github.com/bbye98/mdcraft) package which can easily be installed into a conda environment by cloning the github repo and then pip installing.

**Comment by peastman** [comment]
Thanks!  I'll try that test and see if I can reproduce their numbers.

I believe $\gamma$ is the friction, not inverse friction.  See equation 4 of the Groot and Warren paper.  The drag force is proportional to it, not inversely proportional.

**Comment by svarner9** [comment]
Yes I understand that, but in the Trotter algorithm paper, $$\gamma$$ shows up in the exponential term as $$\exp\left(-2\Delta t m \omega^2 / \gamma\right)$$. But in the code, the $$\text{friction}$$ is in the numerator within the exponential. So to me it seems like the $$\text{friction}$$ is inversely proportional to $$\gamma$$? I might be understanding that incorrectly though.

**Comment by peastman** [comment]
There's clearly something wrong in that paper.  You can see from eq. 8 that $B =  2 \gamma/m \omega^2$ has units of inverse time.  That means the term in the exponential has units of time squared, which is clearly wrong.

**Comment by svarner9** [comment]
Ah okay that makes more sense then. That's a terrible place for them to have an error.

I ran the simulation with $$\text{friction}=4.5/\tau$$ and $$\Delta t = 0.04\tau$$. The temperature fluctuates around 300K, but still randomly results in a NaN energy error. I haven't been able to determine the source. Have you experienced this at all? The error does not happen immediately, and it does not happen at the same timestep each time either.

The conservative force is very soft and also does not diverge at any point, so I am confused how a NaN could even arise. Also note that there are no bonds in the system.

With a mixed or single precision CUDA platform, I get the following error after a random amount of time. Temperature is the fourth value from the end, and the potential energy is the second value from the left.
```ruby
82430,272659.2718088096,89356.03319001422,362015.3049988238,298.52945969552786,7999.999999999996,1.93e+03,0:24
82440,272207.8670067012,90449.01280527552,362656.8798119767,302.18099392722695,7999.999999999996,1.93e+03,0:24
82450,273011.98497052555,90862.05934879377,363874.0443193193,303.56094060864916,7999.999999999996,1.93e+03,0:24
82460,273364.9500355532,1.3835058063020812e+19,1.3835058063021085e+19,4.6221528205343576e+16,7999.999999999996,1.93e+03,0:24
82470,273125.4320111368,2.8360424495267238e+19,2.8360424495267512e+19,9.474930677936683e+16,7999.999999999996,1.93e+03,0:24
Traceback (most recent call last):
  File "/home/svarner/OpenMM-DPD/simple/sim.py", line 203, in <module>
    md(N, ts,
  File "/home/svarner/OpenMM-DPD/simple/sim.py", line 172, in md
    simulation.step(every)
  File "/home/svarner/miniconda3/envs/test/lib/python3.11/site-packages/openmm/app/simulation.py", line 150, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/home/svarner/miniconda3/envs/test/lib/python3.11/site-packages/openmm/app/simulation.py", line 263, in _simulate
    self._generate_reports(unwrapped, False)
  File "/home/svarner/miniconda3/envs/test/lib/python3.11/site-packages/openmm/app/simulation.py", line 282, in _generate_reports
    reporter.report(self, state)
  File "/home/svarner/miniconda3/envs/test/lib/python3.11/site-packages/openmm/app/statedatareporter.py", line 204, in report
    self._checkForErrors(simulation, state)
  File "/home/svarner/miniconda3/envs/test/lib/python3.11/site-packages/openmm/app/statedatareporter.py", line 363, in _checkForErrors
    raise ValueError('Energy is NaN.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan')
ValueError: Energy is NaN.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan
```

With a double precision CUDA platform, the potential energy will randomly rise to another value, and the temperature explodes, but there is no error. See the log below where the temperature jumps around step 51130. Note that the energy also rises to a higher value eventually but it isn't shown in this specific log.
```ruby
50930,272638.5018803381,90658.2596936064,363296.76157394453,302.88006659514133,7999.999999999996,836,2:40
50940,272940.78317362024,90235.30554819417,363176.0887218144,301.4670196189244,7999.999999999996,836,2:40
50950,272622.9958124974,91000.31843566838,363623.31424816576,304.0228502193288,7999.999999999996,836,2:40
50960,272753.18639727955,91033.00381771002,363786.1902149896,304.1320487713737,7999.999999999996,836,2:40
50970,273056.4745255259,90314.84272559975,363371.31725112564,301.7327452755916,7999.999999999996,836,2:40
50980,272771.64985849324,90403.22315385382,363174.87301234703,302.028015337975,7999.999999999996,836,2:40
50990,272680.2597930075,89821.91237413918,362502.1721671467,300.0859148799749,7999.999999999996,836,2:40
51000,272467.541280014,90118.36849929637,362585.90977931034,301.0763447782942,7999.999999999996,836,2:40
51010,272450.5998234421,89745.04304722814,362195.64287067024,299.82910224169325,7999.999999999996,836,2:40
51020,272741.8251299733,90463.06465308763,363204.88978306093,302.2279397280343,7999.999999999996,836,2:40
51030,273081.2046978192,90353.36858853108,363434.5732863503,301.86145628294724,7999.999999999996,836,2:40
51040,272468.96027893916,89996.63884632445,362465.59912526363,300.6696583326979,7999.999999999996,836,2:40
51050,272605.3158255866,90208.06490925413,362813.38073484076,301.376011402308,7999.999999999996,836,2:40
51060,272683.86276679224,90374.91086342624,363058.77363021846,301.9334268422422,7999.999999999996,836,2:40
51070,272473.0218568486,89860.95158378282,362333.9734406314,300.2163409267206,7999.999999999996,836,2:40
51080,272847.6890005629,89044.68267022561,361892.3716707885,297.4892691327847,7999.999999999996,836,2:40
51090,272857.3807030594,90240.8717466304,363098.2524496898,301.4856157243288,7999.999999999996,836,2:40
51100,272911.2833717771,90334.0676570841,363245.3510288612,301.7969738251793,7999.999999999996,836,2:39
51110,272746.09450323007,90434.56874182142,363180.6632450515,302.1327377736711,7999.999999999996,836,2:39
51120,272437.05752696,89384.2672631757,361821.32479013567,298.62378687529394,7999.999999999996,836,2:39
51130,273074.2774096422,1.3835058057154044e+19,1.3835058057154316e+19,4.62215281857433e+16,7999.999999999996,836,2:39
51140,272861.0469121804,1.329422945320789e+19,1.3294229453208162e+19,4.441467457821348e+16,7999.999999999996,836,2:39
51150,272824.0941954225,1.3209401930369798e+19,1.320940193037007e+19,4.413127441309742e+16,7999.999999999996,836,2:39
51160,273280.59791767946,1.3209397523130446e+19,1.3209397523130718e+19,4.413125968895697e+16,7999.999999999996,836,2:39
51170,273067.9553829482,1.3209397512197976e+19,1.3209397512198248e+19,4.4131259652432696e+16,7999.999999999996,836,2:39
51180,273042.63504679734,1.3209397257866697e+19,1.320939725786697e+19,4.413125880273764e+16,7999.999999999996,836,2:39
51190,272975.5296756814,1.3209384678306875e+19,1.3209384678307148e+19,4.4131216775701944e+16,7999.999999999996,836,2:39
51200,272974.8272794363,1.3209384676573893e+19,1.3209384676574165e+19,4.413121676991222e+16,7999.999999999996,836,2:39
51210,273400.47882803634,1.320936640253809e+19,1.3209366402538363e+19,4.413115571820883e+16,7999.999999999996,836,2:39
51220,273556.0054976625,1.320935808225463e+19,1.3209358082254905e+19,4.4131127920984216e+16,7999.999999999996,836,2:39
51230,273673.0387733737,1.3209358080329013e+19,1.3209358080329288e+19,4.413112791455092e+16,7999.999999999996,836,2:39
51240,274083.8482458485,1.3209337929076822e+19,1.3209337929077096e+19,4.413106059125764e+16,7999.999999999996,836,2:39
```

I visualized the simulation and looked at the radial distribution function before and after the jump. Before the jump, the $$g(r)$$ as the expected behavior.

![Image](https://github.com/user-attachments/assets/8855526f-b903-434b-94de-99512b2ce0b0)

While, after the jump, it looks more like a gas.

![Image](https://github.com/user-attachments/assets/f488fec5-26a1-49a3-a0b1-a8fe7ea696d0)

I am wondering if it could have something to do with random number generation. In the Groot and Warren paper, they mention that they get the same statistics by using a uniform distribution with the same variance, so they elect to use that instead. Should we also consider using a uniform distribution?

**Comment by peastman** [comment]
I'm trying to create a unit test that will do the check you suggested.  Here's what I have so far.  It creates a box of particles, runs a short simulation, and prints the diffusion coefficient.

```c++
void testHydrodynamics() {
    const int gridWidth = 15;
    const int numParticles = gridWidth*gridWidth*gridWidth;
    const double density = 3.0;
    const double boxSize = pow(numParticles/density, 1.0/3.0);
    const double cutoff = 1.0;
    const double friction = 6.75;
    const double dt = 0.04;
    const double temp = 1.0/BOLTZ; // So kT = 1.0 kJ/mol
    System system;
    system.setDefaultPeriodicBoxVectors(Vec3(boxSize, 0, 0), Vec3(0, boxSize, 0), Vec3(0, 0, boxSize));
    CustomNonbondedForce* force = new CustomNonbondedForce("15*(1-r)");
    force->setNonbondedMethod(CustomNonbondedForce::CutoffPeriodic);
    for (int i = 0; i < numParticles; i++) {
        system.addParticle(1.0);
        force->addParticle();
    }
    system.addForce(force);
    vector<Vec3> positions(numParticles);
    OpenMM_SFMT::SFMT sfmt;
    init_gen_rand(0, sfmt);
    double scale = boxSize/gridWidth;
    for (int i = 0; i < gridWidth; i++)
        for (int j = 0; j < gridWidth; j++)
            for (int k = 0; k < gridWidth; k++)
                positions[i*gridWidth*gridWidth + j*gridWidth + k] = scale*Vec3(i+0.1*genrand_real2(sfmt), j+0.1*genrand_real2(sfmt), k+0.1*genrand_real2(sfmt));
    DPDIntegrator integrator(temp, friction, cutoff, dt);
    Context context(system, integrator, platform);
    context.setPositions(positions);
    context.setVelocitiesToTemperature(temp, 0);
    integrator.step(100);
    State state1 = context.getState(State::Positions);
    integrator.step(1000);
    State state2 = context.getState(State::Positions);
    double r2 = 0.0;
    for (int i = 0; i < numParticles; i++) {
        Vec3 delta = state1.getPositions()[i]-state2.getPositions()[i];
        r2 += delta.dot(delta);
    }
    r2 /= numParticles;
    printf("%g\n", r2/6/(state2.getTime()-state1.getTime()));
}
```

In the paper they come up with an analytical estimate of 0.354 for it.  Then they say they computed it from a simulation, but don't say how, and came up with 0.306, which they consider accurate enough.  My estimate is off by about the same amount in the opposite direction, around 0.42.  They don't say what value they used for the repulsion parameter.  Their analytical estimate predicts it shouldn't matter, but in my tests it does matter.

| a | D |
|---|---|
10|0.32
15|0.42
20|0.69

> they mention that they get the same statistics by using a uniform distribution with the same variance, so they elect to use that instead.

That seems like a questionable choice.  If you use uniform random numbers, you aren't integrating a Langevin equation.  They don't offer any justification for it except that it gives the same average temperature.  But it could easily affect other properties.

**Comment by peastman** [comment]
I looked at their derivation of the estimate for the diffusion coefficient.  It begins (emphasis added), "Focus on the equation of motion of a single particle and *ignore the conservative forces*..."  No wonder it doesn't depend on the repulsion parameter!  They explicitly chose to ignore any dependence on it.

Now look at section VI, which is where they make use of the estimate.  They state,

> In a typical fluid, water for instance, Sc is of order $10^3$, reflecting the fact that momentum is transported more efficiently than particles, as a consequence of the caging effect of the interparticle potential. Since in DPD, very soft potentials are used, this caging effect is expected to be reduced, and one might expect Sc to be reduced. A simple calculation given below indicate that this is indeed the case.

Consider what they just said there.  They expect the value of Sc to be reduced because of the soft interparticle potential, and they will demonstrate it's true with a calculation that *completely ignores the interparticle potential*, and hence provides no information at all about the effect of the potential in DPD.

I'm really not very impressed with this paper. :(

**Comment by peastman** [comment]
I think your problem is probably in these values:

```python
ROOM_TEMP = 300.0 * unit.kelvin        # Room temperature
MW = 1.0 * unit.amu                     # Water molar mass
DIAMETER = 1.0 * unit.nanometer         # Water molecule size
```

If you're trying to use physical units, the mass of a water molecule is much more than 1 amu, to say nothing of the cluster of water molecules each particle represents.  On the other hand, if you're using the reduced units conventionally used in DPD (mass=1, cutoff=1, kT=1), then your temperature is much too high.  You want to set it to `1*kilojoules_per_mole/MOLAR_GAS_CONSTANT_R`, which is about 120.

**Comment by svarner9** [comment]
Hm, yeah that is odd. If I may try an interpret what they are saying... I think what they mean is that there is no caging in DPD, and so the diffusion is largely determined by the drag and random forces as the particles inevitably pass by each other at a given average velocity, which is why it may suffice to ignore the conservative pair potential in their simple derivation. In normal MD where caging occurs, then the diffusion can be largely determined by the pair potential, and the friction/drag may not be dominant. In addition, although they consider a single particle, they do assume a uniform density of the surrounding fluid $$g(r)=1$$ to evaluate the drag and random forces.

When they compute $$D$$ and $$\nu$$ from the simulation, I guess they don't use a conservative force? If they do, I would assume they use $$a=25$$ which is the standard for $$\rho=3$$. In general, the accepted convention is to use $$a/k_B T=75/\rho$$.

I also wanted to update that I only experienced the huge temperature rise on the CUDA platform. I just ran a few (smaller) simulations on the CPU platform and they did not have this issue. There is a small caveat that it may take longer to observe this seemingly stochastic error in a system with much less particles, particularly if it has to do with the pairwise drag or random force.

Does the actual value of the temperature matter? I solve for the simulation timescale as $$\tau=\sqrt{M_W r_c^2/k_B T}$$ and scale the timestep and friction based on that. Thus, if I decrease the temperature then it will just result in a larger $$\tau$$, but it would be my understanding that the simulation should be identical. Unless there is somewhere in the integrator that $$k_B T$$ or $$M_W$$ arise independently without multiplication by either $$\gamma$$ or $$\Delta t$$.

**Comment by svarner9** [comment]
I just tried decreasing $$T$$ down to 120 K and the same NaN energy occurs.

I tried running with no conservative force and the error disappears, likely due to the fact that $$U=0$$. The temperature is very stable, albeit not necessarily at the correct value.

If I run without a conservative force then the equilibrium temperature depends heavily on the timestep (see figure below). I am not sure if this should be the case or not. Please let me know if this is normal and expected behavior. It was my understanding that the random and drag forces should set the temperature irrespective of the conservative force used. Maybe this has to do with the discussion below eq. 10 in Groot and Warren, and the fact that the random force has to depend on $$\Delta t^(-1/2)$$. I am sure the Trotter algorithm deals with this, but it is not obvious to me exactly how.

![Image](https://github.com/user-attachments/assets/2387ea73-84ff-4201-aaae-1b226ea581c7)

**Comment by svarner9** [comment]
I tested in LAMMPS with the conservative force turned off, and the temperature is indeed independent of the timestep. That is to say, as long as the timestep is not extremely large, the reduced temperature always fluctuates around 1. So, that leads me to believe that there could be something off in the current implementation.

**Comment by peastman** [comment]
Agreed, that doesn't sound right.  There also shouldn't be any difference between CUDA and CPU.  Does the temperature varying with step size happen on both?

**Comment by svarner9** [comment]
The result when using the CPU implementation is different than the CUDA one (see figure below). In addition, I noticed that the equilibrium temperature that is achieved also heavily depends on the number of particles in the system, and also on the initial state. For the same exact system, the equilibrium temperature is different across several runs, especially for relatively small systems. This randomness becomes less pronounced for larger systems.

As a singular example, with parameters $$\rho=3$$, $$T=120$$ K, $$\gamma=4.5$$, $$\Delta t=0.05$$, and $$a=0$$, I ran a few simulations and extracted the equilibrium temperature. For $$N=3000$$ I get $$T\approx 128$$ K, for $$N=24000$$ I get $$T\approx 151$$ K, and for $$N=81000$$ I get $$T\approx 165$$ K. Note that none of these are very close to the target of 120 K.

![Image](https://github.com/user-attachments/assets/905d98d7-8199-433e-ba5a-49fab1f654cc)

**Comment by peastman** [comment]
> If I run without a conservative force then the equilibrium temperature depends heavily on the timestep

The force is essential, because that's how it knows whether to apply periodic boundary conditions.  If there's no force, it concludes the system is non-periodic.  All the particles will just spread out until they stop interacting with each other, and then the thermostat won't do anything.

**Comment by svarner9** [comment]
Ah okay I see! So I can use the conservative force and just set $$a=0$$ to get the same effect without interfering with the periodic boundary conditions.

**Comment by peastman** [comment]
Right.

**Comment by peastman** [comment]
I can reproduce the NaN when running on an NVIDIA GPU.  It happens with both CUDA and OpenCL platforms.  I can't reproduce it using OpenCL on an Apple GPU.  That suggests it might be something hardware specific...

**Comment by peastman** [comment]
I found the cause of the nan.  It happened when the random number generator produced 0, which happened every 4 billion interactions.  It's fixed now.

**Comment by svarner9** [comment]
Great! That sounds like a very hard error to debug. I am glad that you were able to find it. I'll do some more testing next week.

**Comment by peastman** [comment]
I think something is still up with the GPU version.  It gives much larger diffusion coefficients than the reference version.  You might want to hold off on testing until I track it down.

**Comment by svarner9** [comment]
Okay sounds good! I will until that is resolved.

**Comment by peastman** [comment]
I think I have the problems sorted out.  CPU and GPU platforms now give the same diffusion coefficient, though I'm still searching for a good reference value to compare to.  Go ahead and test!

**Comment by peastman** [comment]
I'm trying to get a test for diffusion coefficient working.  It's very confusing trying to match up different papers that all define the algorithm and parameters differently.  For a system with no conservative force, my test produces a diffusion coefficient of about 0.8, compared to about 0.35 predicted by the estimate in Groot and Warren.  But I'm not sure if I've really matched them up correctly.

Most papers define $\omega(r)$ as $1-r/r_c$.  But in Español and Warren they instead define it as $2(1-r/r_c)$ "in order to remain close to the original algorithm" in Hoogerbrugge and Koelman.  If I use that form I get pretty good agreement.  But I'm not sure whether that's just a coincidence, or whether it's supposed to compensate for some other difference.

**Comment by peastman** [comment]
The above is with friction=6.75, density=3.0, dt=0.04.  Changing dt has little effect on diffusion, which is consistent with the prediction.  Changing density does affect it, and the simulation result continues to match the predicted value well.  But if I change the friction, it no longer matches well.

**Comment by svarner9** [comment]
So your diffusion coefficient is inversely proportional to density, but does not follow the same trend with respect to $$\gamma$$?

Yeah these subtleties seem very hard to iron out... I can compute the diffusion coefficient in LAMMPS for various $$\rho$$ and $$\gamma$$ if that would be helpful. They are pretty explicit about what equations they use, so there should be no confusion.

![Image](https://github.com/user-attachments/assets/d59e3308-a903-4057-a969-942af10915be)

**Comment by peastman** [comment]
That would be great.  We can see how well they agree.  I'm doing this with no conservative force (A=0), which should make the analytical estimate more accurate.

**Comment by svarner9** [comment]
Sorry not yet, things have been very busy with APS coming up. I will do that today though and let you know!

**Comment by svarner9** [comment]
I ran LAMMPS simulations with $$A=0$$ and $$r_c=1$$ for various $$\rho$$ and $$\gamma$$.

My results follow the correct trends, but do not match up with the exact numbers reported in the Groot and Warren paper. As an example, I get $$D=0.5$$ for $$\rho=3$$ and $$\gamma=6.75$$, but they reported $$D=0.3$$. I am not sure where the difference is from. I am curious if your results align well with mine or not (I assume not since you said you got a value of 0.8 for the example case).

![Image](https://github.com/user-attachments/assets/024fd356-8a12-48dc-aece-08e823d416f2)

Interestingly, it seems that my simulations obey the expected result less and less as $$\gamma$$ becomes larger. See the plot below. In addition, my simulation result is always higher than the predicted result from the theory, whereas the Groot and Warren paper has a lower value (0.3) from the simulation when compared to the theory (0.35).

![Image](https://github.com/user-attachments/assets/fe732ca7-2f89-4097-81dc-460ee4a31697)

**Comment by peastman** [comment]
Thanks!  Can you post the actual numbers so I can compare my results to them directly?

**Comment by svarner9** [comment]
```
rho gamma D
2.00 0.50 7.4086
2.00 1.00 3.8515
2.00 2.00 2.0524
2.00 3.00 1.4654
2.00 4.50 1.0466
2.00 6.75 0.7609
3.00 0.50 4.9356
3.00 1.00 2.5876
3.00 2.00 1.3786
3.00 3.00 0.9764
3.00 4.50 0.6901
3.00 6.75 0.5203
4.00 0.50 3.7593
4.00 1.00 1.9350
4.00 2.00 1.0302
4.00 3.00 0.7327
4.00 4.50 0.5216
4.00 6.75 0.3791
5.00 0.50 3.0185
5.00 1.00 1.4981
5.00 2.00 0.8048
5.00 3.00 0.5656
5.00 4.50 0.4147
5.00 6.75 0.3018
```

**Comment by peastman** [comment]
Thanks!  What is the value of dt?

**Comment by svarner9** [comment]
I used 0.04, but that shouldn't matter right? My $$D$$ is with respect to time (steps*dt), not number of steps.

**Comment by peastman** [comment]
The theoretical estimate doesn't depend on the step size, but in practice it can.  See figure 1c in https://doi.org/10.1209/0295-5075/79/36002.

**Comment by peastman** [comment]
I get excellent agreement with LAMMPS if I multiply $\omega(r)$ by $\sqrt{2}$.  That's encouraging.  I just need to figure out where I missed a factor of 2...

**Comment by peastman** [comment]
I added a test case that computes the diffusion coefficient for $\rho=3$, $\gamma=6.5$ and compares it to your value from LAMMPS.  Here's what I get for the whole set of values.

|Density|Friction|LAMMPS|OpenMM|Ratio|
|---|---|---|---|---|
2|0.5|7.4086|7.09829|0.958114893502146
2|1|3.8515|3.86438|1.00334415162924
2|2|2.0524|2.06065|1.00401968427207
2|3|1.4654|1.50761|1.02880442200082
2|4.5|1.0466|1.02368|0.97810051595643
2|6.75|0.7609|0.812315|1.06757129714811
3|0.5|4.9356|4.77474|0.967408217845855
3|1|2.5876|2.59975|1.00469547070645
3|2|1.3786|1.3726|0.995647758595677
3|3|0.9764|0.9782|1.00184350675952
3|4.5|0.6901|0.729602|1.05724097956818
3|6.75|0.5203|0.548959|1.05508168364405
4|0.5|3.7593|3.57464|0.950879153033809
4|1|1.935|1.94086|1.00302842377261
4|2|1.0302|1.0768|1.04523393515822
4|3|0.7327|0.757938|1.03444520267504
4|4.5|0.5216|0.535076|1.02583588957055
4|6.75|0.3791|0.432074|1.1397362173569
5|0.5|3.0185|2.95471|0.97886698691403
5|1|1.4981|1.54308|1.03002469795074
5|2|0.8048|0.843494|1.04807902584493
5|3|0.5656|0.587647|1.03897984441301
5|4.5|0.4147|0.432102|1.04196286472149
5|6.75|0.3018|0.327583|1.08543074884029

At this point, I think we're just waiting on you to try it out.

**Comment by svarner9** [comment]
> I added a test case that computes the diffusion coefficient for ρ=3, γ=6.5 and compares it to your value from LAMMPS. Here's what I get for the whole set of values.

Did you use  $$\gamma=6.5$$ or $$\gamma=6.75$$? My calculations were for $$\gamma=6.75$$. Just double-checking that we are actually getting agreement for the same simulation parameters, and also double-checking that the test uses the correct value.

Did you end up finding the factor of 1/2 that made the difference?

**Comment by peastman** [comment]
@svarner9 are you still planning to test it?  We can't merge it until it's tested.
